### ISSUE TITLE
Apple login for Firebase

### ISSUE THREAD

**Issue by ColorTwist** [firebase/quickstart-unity#536]
Since Apple currently **doesn't** approve new apps that have third party loggin providers (facebook, Gmail, etc..)  without  Apple Login.

Any estimated when it will be added to Firebase Unity SDK?

**Comment by DellaBitta** [comment]
Hi @ColorTwist,

yes, this is currently on our feature roadmap, though I cannot comment on specific release timelines.

**Comment by ColorTwist** [comment]
Unity firebase SDK 6.9.0:
"(iOS): Added support for linking Apple Sign-in credentials."
It's that, right?
Any example of how to use it?

**Comment by joses1234** [comment]
I am also interested in this, can you please provide/update firebase documentations for this? Thanks!

**Comment by DellaBitta** [comment]
Hi,

Unfortunately there was a mixup in the 6.9.0 release and the feature isn't available in those libraries.  We're working on getting it out soon.

**Comment by joses1234** [comment]
@DellaBitta any estimates on a new release with apple sign in? It's becoming a problem as mentioned by OP apple doesn't approve new apps with other third party login providers.

Are there any alternatives in the meantime? Thank you

**Comment by DellaBitta** [comment]
We're currently validating the next release with the included change.  I cannot comment on specific release dates but it should be available quite soon.  There is no clear alternative at this time. 

As a note, the change adds a new overloaded method to OAuthProvider GetCredential which takes a valid Apple SignIn token and the nonce used to create it.  The change will still require the use of a third party library to do the Apple Sign-in beforehand.

**Comment by joses1234** [comment]
@DellaBitta Thanks for the info!

**Comment by phedg1** [comment]
@DellaBitta Your comment that "The change will still require the use of a third party library to do the Apple Sign-in beforehand." concerns me. For native Android, Firebase seems handle almost all of the Sign in with Apple flow. [https://firebase.google.com/docs/auth/android/apple](url) Is this not going to be the case for Unity?

I have been trying for weeks to get Sign in with Apple to function on Android in a Unity build. I have used Firebase to host my apple-app-site-association file. I can use REST to get a user to sign in, from which Apple returns a security code, to later be exchanged for the token. Apple will do a HTTP POST to the redirect_uri containing the code. This redirect_uri cannot be localhost or 127.0.0.1, like it can be for other OAuth providers. If it could I could simply listen in the app and catch the code as it arrived. Without it I haven't the foggiest how I can pass that code on to app client which requested it. This has been causing me a lot of stress, but I figured that the Firebase solution would handle this, like it does on its other platforms. Please tell me I'm wrong and that I have misunderstood your reply? Please?

**Comment by DellaBitta** [comment]
Hi @phedg1,

There are some 3rd-party plugins to facilitate sign-in to Apple itself.  One is offered by Unity via the asset store ([it has Android support](https://blogs.unity3d.com/2019/09/19/support-for-apple-sign-in/)), some others are hosted on github.  These could be used to faciliate the generation of an AppleID token by signing into Apple.  This token, along with the nonce that was provided during Apple Sign-In, can then be fed into the OAuthProvider::GetCredential method for FirebaseAuth for projects which have Apple enabled as a sign in provider.

I hope this helps!

**Comment by phedg1** [comment]
@DellaBitta I am saddened that you have confirmed that Firebase won't handle the Apple OAuth for Unity developers, as it is handled on Firebase's other platforms. Apple's implementation makes it significantly more difficult than other OAuth providers. 

The Unity asset store offering only works on iOS. The article didn't mention it having Android functionality at all, nor can I find any other information elsewhere that says it does. It did say that Unity was looking at rolling out an Authentication package in the future, but that has not been released into alpha yet.

I will try look for the GitHub OAuth solutions. Thank you.

**Comment by DellaBitta** [comment]
Hi @phedg1,

I spoke with the Auth team, and on Android devices you may use the Generic IDP Sign-In flow.  On Android this uses the default UIActivity to capture user sign-in information.

We're working on publishing the documentation on how to use this feature in Unity. In the meantime please check the [Android Apple Sign-In guide](https://firebase.google.com/docs/auth/android/apple) on how to configure your project in the Firebase Console. 

Example Unity code would look like this:

```// Initialize Auth as you normally would.
Firebase.Auth.FirebaseAuth auth = Firebase.Auth.FirebaseAuth.DefaultInstance;

Firebase.Auth.FederatedOAuthProviderData data = new              
        Firebase.Auth.FederatedOAuthProviderData();      

data.ProviderId = “apple.com”;

// configure oauth scopes
List<string> scopes = new List<string>();
scopes.Add(scope1);
scopes.Add(…);
data.Scopes = scopes;

// configure oauth custom parameters
data.CustomParameters = new Dictionary<string, string>();
data.CustomParameters.Add(customParameterKey1, customParameterValue1);
data.CustomParameters.Add(…, …);

// create provider object
Firebase.Auth.FederatedOAuthProvider provider = new Firebase.Auth.FederatedOAuthProvider(data);

// sign in with provider object
auth.SignInWithProviderAsync(provider).ContinueOnMainThread(task => {
        if (task.IsCanceled) {
            Debug.LogError("SignInWithProviderAsync was canceled.");
            return;
        }
        if (task.IsFaulted) {
            Debug.LogError("SignInWithProviderAsync encountered an error: " + task.Exception);
            return;
        }

        Firebase.Auth.SignInResult signInResult = task.Result;
        Friebase.Auth.FirebaseUser user = signInResult.User;
        Debug.LogFormat("User signed in successfully: {0} ({1})",
            user.DisplayName, user.UserId);
    });
```
This code has been released and should work today on Android.   However, please note that this will not work on iOS. The aforementioned and upcoming OAuthProvider GetCredential (with nonce) flow must be used on that platform.

**Comment by shlighter** [comment]
I'm using Unity's SignInwIthApple asset, and it seems to log in fine on ios.  
Since firebase doesn't support SignInwIthApple,  I need to think about an idea of how to convert the logged in user to be also on firebase
For example, in my SignInwIthApple script, after a successful log in, I try to take the user's apple email and save it on firebase using Email and password (create it for them behind the scenes)
the problem is that  args.userInfo.email returns an empty string .
But I can use 2 other identification after the login, I can get args.userInfo.userId which is a unique aplpe ID that I can save in firebase, the problem is that it's too long, so can't use it as an email,
I can also get args.token which gives me a long token
How would you recommend me to save the user in Firebase? My apple sign in works fine, I just need to take the user unique details and store in firebase for sign up and for future log in, apple doesn't approve my app without apple sign in

**Comment by phedg1** [comment]
@shlighter , DellaBita's comment above says that you will be able to use an Apple auth token to authenticate in Firebase in the _next_ release Firebase makes for Unity. This functionality is coming really soon but it's not available right at this moment. https://github.com/firebase/quickstart-unity/issues/536#issuecomment-574869158

**Comment by phedg1** [comment]
Ladies and gentlemen, the update has been published!

**Comment by a-maurice** [comment]
Yep, Firebase Unity 6.10.0 has been released which has the needed OAuthProvider GetCredential (with nonce) flow.

It can be downloaded from https://firebase.google.com/download/unity

**Comment by HarrisonShort** [comment]
> Hi @phedg1,
> 
> I spoke with the Auth team, and on Android devices you may use the Generic IDP Sign-In flow. On Android this uses the default UIActivity to capture user sign-in information.
> 
> We're working on publishing the documentation on how to use this feature in Unity. In the meantime please check the [Android Apple Sign-In guide](https://firebase.google.com/docs/auth/android/apple) on how to configure your project in the Firebase Console, and please use this code as an example:
> 
> ```
> Firebase.Auth.FirebaseAuth auth = Firebase.Auth.FirebaseAuth.DefaultInstance;
> 
> Firebase.Auth.FederatedOAuthProviderData data = new              
>         Firebase.Auth.FederatedOAuthProviderData();      
> 
> data.ProviderId = “apple.com”;
> 
> // configure oauth scopes.
> List<string> scopes = new List<string>();
> scopes.Add(scope1);
> scopes.Add(…);
> data.Scopes = scopes;
> 
> // configure oauth custom parameters.
> data.CustomParameters = new Dictionary<string, string>();
> data.CustomParameters.Add(customParameterKey1, customParameterValue1);
> data.CustomParameters.Add(…, …);
> 
> // create provider object.
> Firebase.Auth.FederatedOAuthProvider provider = new Firebase.Auth.FederatedOAuthProvider(data);
> 
> // sign in with provider object.
> auth.SignInWithProviderAsync(provider).ContinueOnMainThread(task => {
>         if (task.IsCanceled) {
>             Debug.LogError("SignInWithProviderAsync was canceled.");
>             return;
>         }
>         if (task.IsFaulted) {
>             Debug.LogError("SignInWithProviderAsync encountered an error: " + task.Exception);
>             return;
>         }
> 
>         Firebase.Auth.SignInResult signInResult = task.Result;
>         Friebase.Auth.FirebaseUser user = signInResult.User;
>         Debug.LogFormat("User signed in successfully: {0} ({1})",
>             user.DisplayName, user.UserId);
>     });
> ```
> 
> The SignInWithProvider functionality was released in 6.9.0 and should work today on Android. However, please note that this will not work on iOS. The aforementioned and upcoming OAuthProvider GetCredential (with nonce) flow must be used on that platform.
Firstly, thanks for all your hard work on Firebase, it's made application development through Unity so much simpler. 
I'm currently setting up the Android Apple Sign-Up flow, and have been using the above snippet (though initially through this [page](https://firebase.google.com/docs/auth/unity/apple?authuser=0)). I'm not sure whether I'm missing something, but "auth.SignInWithProviderAsync(provider).ContinueOnMainThread" doesn't seem to actually exist.
